### Part 2: Data Science

In [2]:
import os
import numpy as np
import pandas as pd

In [3]:
os.chdir('C:\\Users\\akash\\Desktop\\AppScrip\\New folder')

In [70]:
train = pd.read_csv('sample_train.csv', header = 0)
test = pd.read_csv('sample_test.csv', header = 0)
train = train.drop('Unnamed: 0', axis =1)
test = test.drop('Unnamed: 0', axis =1)

In [74]:
print('Train Shape',train.shape)
print("test Shape",test.shape)

Train Shape (169, 8)
test Shape (50, 8)


All the features are binary categorical data set
so the possible combination are 2^8 = 256

In [72]:
# Creating the dummy numpy array similar to the dimension of the dataset
dummy_train = pd.DataFrame(np.random.randint(2, size = (250,8)), columns = train.columns.values)
dummy_test = pd.DataFrame(np.random.randint(2, size = (50,8)), columns= train.columns.values)
#Concatination the dataset by dropping the dumplicate values
train = pd.concat([train,dummy_train]).drop_duplicates().reset_index(drop = True)
test = pd.concat([test,dummy_test]).drop_duplicates().reset_index(drop = True)
# Creating the balance target value with balance data set
target_val = np.array(train[train.columns.values[4]])
np.random.shuffle(target_val)
train['target'] = target_val
target_test = np.array(test[test.columns.values[-2]])
np.random.shuffle(target_test)
test['target'] = target_test

In [85]:
from sklearn import tree

In [ ]:
x_train = train.drop('target', axis =1)
y_train = train['target']
x_test = test.drop('target', axis = 1)
y_test = test['target']

In [95]:
dtree_gini = tree.DecisionTreeClassifier(random_state= 42).fit(x_train,y_train)# Default Gini Index
dtree_entropy = tree.DecisionTreeClassifier(criterion='entropy', random_state= 42).fit(x_train,y_train)

In [96]:
gini_predict = dtree_gini.predict(x_test)
entropy_predict = dtree_entropy.predict(x_test)

In [104]:
with open('gini_tree.txt', 'w') as f:
    f = tree.export_graphviz(dtree_gini, out_file=f)

In [105]:
with open('entropy_tree.txt', 'w') as f:
    f = tree.export_graphviz(dtree_entropy, out_file=f)

In [116]:
def error_metrics(y_true,y_predict):
    confuse = confusion_matrix(y_true, y_predict)
    print(confuse)
    tn, fp, fn, tp = confuse.ravel()
    print('Accurary =', (tn+tp)/(tn+fp+fn+tp))
    print('Error = ', 1- ((tn+tp)/(tn+fp+fn+tp)))
    print("TPR =",tp/(fp+tp))
    print("TNR =",tn/(tn+fp))
    print("FPR =",fp/(tn+fp))
    print("FNR =",fn/(fp+tp))
    print('Precision =', tp/(tp+fp))
    print('Recall =', tp/(tp+fn))
    from sklearn.metrics import f1_score
    print('F1_score :', f1_score(y_true,y_predict))

In [92]:
from sklearn.metrics import confusion_matrix

In [117]:
#GINI CONFUSION METRIX
confusion_matrix(y_test, gini_predict)
error_metrics(y_test, gini_predict)

[[11 15]
 [ 9 15]]
Accurary = 0.52
Error =  0.48
TPR = 0.5
TNR = 0.4230769230769231
FPR = 0.5769230769230769
FNR = 0.3
Precision = 0.5
Recall = 0.625
F1_score : 0.5555555555555556


In [118]:
#Entropy CONFUSION METRIX
confusion_matrix(y_test, entropy_predict)
error_metrics(y_test, entropy_predict)

[[11 15]
 [ 9 15]]
Accurary = 0.52
Error =  0.48
TPR = 0.5
TNR = 0.4230769230769231
FPR = 0.5769230769230769
FNR = 0.3
Precision = 0.5
Recall = 0.625
F1_score : 0.5555555555555556


Performance of both the algorithms are same but the accuracy of an algorithm is 52% where as the precision of an algorithm is 50% and recall rate in 62.5%.
F1-score(Harmonic mean of precision and recall) = 55.555

Detailed explanation of the algorithm is on DecisionTree.docx file